pageserver/src/tenant.rs

```diff
@@ -112,34 +182,511 @@ pub struct Tenant {
+        let mut index_parts: HashMap<TimelineId, &IndexPart> = HashMap::new();
+        for (timeline_id, index_part) in remote_timelines.iter() {
+            let remote_metadata = TimelineMetadata::from_bytes(&index_part.metadata_bytes)
+                .with_context(|| {
+                    format!(
+                        "Failed to parse metadata file from remote storage for tenant {}",
+                        self.tenant_id
+                    )
+                })?;
+            timeline_ancestors.insert(*timeline_id, remote_metadata);
+            index_parts.insert(*timeline_id, index_part);
+        }
+
+        let sorted_timelines = tree_sort_timelines(timeline_ancestors)?;
+
+        let mut attached_timelines = Vec::new();
+
+        // For every timeline, download the metadata file.
+        for (timeline_id, _metadata) in sorted_timelines {
+            info!("downloading timeline {}", timeline_id);
+            let index_part = index_parts.get(&timeline_id).unwrap();
+            tokio::fs::create_dir_all(self.conf.timeline_path(&timeline_id, &self.tenant_id))
+                .await
+                .context("Failed to create new timeline directory")?;
+
+            // Download everything
+            let remote_metadata = TimelineMetadata::from_bytes(&index_part.metadata_bytes)
+                .with_context(|| {
+                    format!(
+                        "Failed to parse metadata file from remote storage for tenant {}",
+                        self.tenant_id
+                    )
+                })?;
+
+            let remote_client = create_remote_timeline_client(
+                remote_storage,
+                self.conf,
+                self.tenant_id,
+                timeline_id,
+            )?;
+
+            let ancestor = if let Some(ancestor_id) = remote_metadata.ancestor_timeline() {
+                let timelines = self.timelines.lock().unwrap();
+                Some(Arc::clone(timelines.get(&ancestor_id)
+                    .ok_or_else(|| anyhow!("cannot find ancestor timeline {ancestor_id} for timeline {timeline_id}"))?))
+            } else {
+                None
+            };
+
+            let timeline = Timeline::new(
+                self.conf,
+                Arc::clone(&self.tenant_conf),
+                &remote_metadata,
+                ancestor,
+                timeline_id,
+                self.tenant_id,
+                Arc::clone(&self.walredo_mgr),
+                remote_metadata.pg_version(),
+                Some(remote_client),
+            );
+
+            // Initialize the layer map, based on all the files we now have on local disk.
+            timeline
+                .load_layer_map(remote_metadata.disk_consistent_lsn())
+                .context("failed to load layermap")?;
+
+            // Download everything from remote storage to local disk
+            timeline
+                .reconcile_with_remote(Some(index_part), true)
```

Comment:
Reviving the discussion from the previous PR: why do we need to download entire timeline locally?
We can on demand download now, so all we need is metadata + some heuristics to spawn background download tasks later?